enhancements/network/egress-qos.md

```diff
@@ -0,0 +1,187 @@
+To enable this, we will use Differentiated Services Code Point (DSCP) which allows us to classify packets by setting a 6-bit field in the IP header, effectively marking the priority of a given packet relative to other packets as "Critical", "High Priority", "Best Effort" and so on.
+
+By introducing a new CRD `EgressQoS`, users could specify a DSCP value for packets originating from pods on a given namespace heading to a specified CIDR.
+The CRD will be Namespaced, with one resource allowed per namespace.
+The resources will be watched by ovn-k, which in turn will configure OVN's [QoS Table](https://man7.org/linux/man-pages/man5/ovn-nb.5.html#QoS_TABLE). 
+
+## Motivation
+
+Telco customers require support for DSCP marking capability for some of their 5G applications, giving some pods precedence over others.
+The QoS markings will be consumed and acted upon by objects outside of the OpenShift cluster to optimize traffic flow throughout their networks.
+
+### Goals
+
+- Provide a mechanism for users to set DSCP on egress traffic coming from specific namespaces.
+
+### Non-Goals
+
+- Ingress QoS.
+
+- Consolidating with current `kubernetes.io/egress-bandwidth` and `kubernetes.io/ingress-bandwidth` annotations.
+Nonetheless, the work done here does not interfere with the current bandwidth QoS mechanism.
+
+- The DSCP marking does not need to be handled or acted upon by OpenShift, just added to selected headers.
+
+- Marking East/West traffic, exposing the DSCP value from the inner packet to the outer geneve packet.
+
+## Proposal
+
+To achieve egress DSCP marking on pods, we introduce a new namespace-scoped CRD `EgressQoS` which allows specifying a set of QoS rules, each has a DSCP value and a destination CIDR. Traffic coming from pods on the namespace heading to each destination CIDR will be marked with the corresponding DSCP value.
+
+### Implementation Details/Notes/Constraints
+
+A new API `EgressQoS` under the `k8s.ovn.org/v1` version will be added to `pkg/crd`.
+
+A new controller in OVN-K will watch `EgressQoS` and `Pod` objects, which will create the relevant QoS objects in OVN and result in the necessary flows to be programmed in OVS. By listing the pods in the namespace these QoS rules will be attached only to the relevant node local switches.
+
+For example, assuming there's a single pod `app1` in namespace `default` on node `node1` and the following `EgressQoS` is created:
+
+```yaml
+kind: EgressQoS
+apiVersion: k8s.ovn.org/v1
+metadata:
+  name: default
+  namespace: default
+spec:
+  egress:
+  - dscp: 46
+    dstCIDR: 0.0.0.0/0
+  - dscp: 30
+    dstCIDR: 1.2.3.4/32
+```
+the equivalent of:
+```bash
+ovn-nbctl qos-add node1 from-lport 1 "ip4.src == <default_ns_address_set> && ip4.dst == 0.0.0.0/0" dscp=46
+ovn-nbctl qos-add node1 from-lport 2 "ip4.src == <default_ns_address_set> && ip4.dst == 1.2.3.4/32" dscp=30
```

Comment:
note this "2" is priority:
[--may-exist] qos-add switch direction priority match [dscp=dscp]

how will the CRD determine the priority to use for these?

Reply:
I thought following what egressFirewall is doing (https://github.com/ovn-org/ovn-kubernetes/blob/a24a233130fb87e07557d8c1579ad52a861fde96/go-controller/pkg/ovn/egressfirewall.go#L341) - relying on the rules of the CRD to be in a descending order priority-wise. The user would have to put them in an order that'll work like having a specific CIDR in a higher priority than a broader one that includes it.

Reply:
that API is very confusing to me. do we want to explore an explicit priority field?

Reply:
we could, but I'm not sure how useful it could be for the user.
with the rules for some namespace being something like :
```
  - dscp: 46
    dstCIDR: 1.2.3.4/32
  - dscp: 40
    dstCIDR: 1.2.3.0/24
  - dscp: 30
    dstCIDR: 0.0.0.0/0
```
would it add any functionality to let the user set some priorities here? would it make any difference if he sets them 3,2,1 or 80,75,5?
looking at a larger set of rules, is it simpler to observe the priority of each rule to understand what takes precedence or rather relying on the order (or some other option)?

Reply:
I think we should just make it work like egress firewall